mapgen/iwyu-mapgen-macos.py

```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env python3
+
+##===--- iwyu-mapgen-macos.py ---------------------------------------------===##
+#
+#                     The LLVM Compiler Infrastructure
+#
+# This file is distributed under the University of Illinois Open Source
+# License. See LICENSE.TXT for details.
+#
+##===----------------------------------------------------------------------===##
+
+""" Generate mappings for macOS libc.
+
+Map headers that either start with an underscore or are architecture-specific.
+"""
+import argparse
+import json
+import sys
+
+from operator import itemgetter
+from pathlib import Path
+
+
+def main(dir):
+    """ Entry point. """
+
+    root = Path(dir)
+
+    paths = set(root.glob('**/*.h'))
+
+    includes = []
+
+    mappings = {
+        'pthread_*': 'pthread.h',
+        'sys/_types/_fd_*': 'sys/select.h',
+        'sys/_types/_in_*': 'netinet/in.h',
+        'sys/_types/_o_*': 'fcntl.h',
+        'sys/_types/_s_*': 'sys/stat.h',
+        'sys/_types/_seek_*': 'stdio.h',
+        'sys/ttycom.h': 'sys/ioctl.h',
+    }
+
+    for p in paths:
+        initial = p
+
+        # xlocale/_stdio.h -> _stdio.h
+        if p.parent.name == 'xlocale':
+            public_path = p.parent.with_name(p.name)
+            if public_path in paths:
+                p = public_path
+
+        # _stdio.h -> stdio.h
+        while p.name.startswith('_'):
+            public_path = p.with_name(p.name[1:])
+            if public_path in paths:
+                p = public_path
+            else:
+                break
+
+        # arm -> machine
+        if p.parent.name in {'arm', 'arm64', 'i386'}:
+            public_path = p.parent.with_name('machine') / p.name
+            if public_path in paths:
+                p = public_path
+
+        # sys/time.h -> time.h
+        # do not map machine/profile.h to profile.h which is a kerberos header
+        if p.parent.name in {
+            'arm', 'arm64' 'i386', 'machine', 'pthread', 'sys'
+        } and p.name != 'profile.h':
+            public_path = p.parent.with_name(p.name)
+            if public_path in paths:
+                p = public_path
+
+        # architecture/byte_order.h -> machine/byte_order.h
+        if p.parent.name == 'architecture':
+            public_path = p.parent.with_name('machine') / p.name
+            if public_path in paths:
+                p = public_path
+
+        # map other headers
+        for glob, repl in mappings.items():
+            if p.relative_to(root).full_match(glob):
+                public_path = root / repl
+                if public_path in paths:
+                    p = public_path
+                break
+
+        # we don't know this one
+        if p.name.startswith('_'):
```

Comment:
Again, maybe a little warning might be nice, so users notice when new headers are overlooked? There's a recent example in the libstdcxx generator, that prints to stderr